services/statestorage/merkle/merkle_forest.go

```diff
@@ -0,0 +1,236 @@
+package merkle
+
+import (
+	"fmt"
+	"github.com/orbs-network/orbs-network-go/crypto/hash"
+	"github.com/orbs-network/orbs-spec/types/go/primitives"
+	"github.com/orbs-network/orbs-spec/types/go/protocol"
+	"github.com/pkg/errors"
+	"strings"
+)
+
+type RootId uint64
+type Proof []*Node
+
+var zeroValueHash = hash.CalcSha256([]byte{})
+
+type Node struct {
+	path     string
+	value    primitives.Sha256
+	branches [256]primitives.MerkleSha256
+}
+
+var emptyNode = &Node{value: zeroValueHash}
+var emptyNodeHash = emptyNode.hash()
+
+func createNode(path string, valueHash primitives.Sha256) *Node {
+	return &Node{
+		path:     path,
+		value:    valueHash,
+		branches: [256]primitives.MerkleSha256{},
+	}
+}
+func (n *Node) hasValue() bool {
+	return !zeroValueHash.Equal(n.value)
+}
+func (n *Node) hash() primitives.MerkleSha256 {
+	serializedNode := fmt.Sprintf("%+v", n)
+	return primitives.MerkleSha256(hash.CalcSha256([]byte(serializedNode)))
+}
+func (n *Node) clone() *Node {
+	newBranches := [256]primitives.MerkleSha256{}
+	copy(newBranches[:], n.branches[:])
+	result := &Node{
+		path:     n.path,
+		value:    n.value, // TODO - copy?
+		branches: newBranches,
+	}
+	for k, v := range n.branches {
+		result.branches[k] = v // TODO - copy?
+	}
+	return result
+}
+func (n *Node) hasChildren() bool {
+	for _, v := range n.branches {
+		if len(v) != 0 {
+			return true
+		}
+	}
+	return false
+}
+
+type Forest struct {
+	roots   map[RootId]primitives.MerkleSha256
+	nodes   map[string]*Node
+	topRoot RootId
+}
+
+func NewForest() *Forest {
+	return &Forest{
+		roots: map[RootId]primitives.MerkleSha256{0: emptyNodeHash},
+		nodes: map[string]*Node{emptyNodeHash.KeyForMap(): emptyNode},
+	}
+}
+
+func (f *Forest) GetRoot(height RootId) (primitives.MerkleSha256, error) {
```

Comment:
To someone unfamiliar with the exact terminology of merkle trees, it's very difficult to understand the API.. maybe a comment above every function like `GetRoot` that explains what it does with an example will help somebody understand merkle trees from this code